Better support for PA with empty mesh partitions

pazner opened the pull request:
Building on work by @v-dobrev, this PR improves support for PA in the case where there are empty mesh partitions.
<!--GHEX{"author":"pazner","assignment":"2024-05-11T15:29:38","editor":"tzanio","id":4260,"reviewers":["camierjs","mlstowell","rw-anderson"],"merge":"2025-01-08T01:13:29","approval":"2025-01-04T04:11:53"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4260](https://github.com/mfem/mfem/pull/4260) | @pazner | @tzanio | @camierjs + @mlstowell + @rw-anderson | 5/11/24 | 1/3/25 | 1/7/25 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

pazner commented:
By the way, why does `FiniteElementSpace::GetTraceElement` take an index `i`?

https://github.com/mfem/mfem/blob/87290fbf5674a0c9ab9164fa7b5d30364a164a54/fem/fespace.hpp#L1186

It is never used

https://github.com/mfem/mfem/blob/13144287705c1338561106b62e75a780d15c8751/fem/fespace.cpp#L3276-L3280

pazner commented:
> By the way, why does `FiniteElementSpace::GetTraceElement` take an index `i`?

It seems the reason is to support variable-order spaces, but this was not implemented yet. Should be fixed in c768c5f579861009cc3fe9a11e2679a9d81daa13 and 3b6c6ce6fb7419587770a391d8883b441e91fa9e.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

tzanio commented:
@mlstowell and @rw-anderson, can you please review?

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -3293,7 +3293,7 @@ class DGDiffusionIntegrator : public BilinearFormIntegrator
                                        Vector &y, Vector &dydn) const override;
 
-   const IntegrationRule &GetRule(int order, FaceElementTransformations &T);
+   const IntegrationRule &GetRule(int order, Geometry::Type geom);
```
I think it is better to not change this -- keep it consistent with other `GetRule` methods.

pazner replied:
Should we implement `Mesh::GetTypicalFaceElementTransformations` or something similar?

The other one of this form is `DGTraceIntegrator::GetRule`. Looking at the PA code for that integrator, I see there is actually still a call to `GetFaceElementTransformations(0)`, so that code should be changed as well.

(The reason I made this change is because something like `GetTypicalFaceElementTransformations` was a little tricky).

v-dobrev replied:
It seems like defining `GetTypical*` for transformations is taking things too far. Instead of constructing the transformation, it is simpler to call `GetRule` conditionally if there are actual elements in the local mesh and otherwise use an empty `IntegrationRule`. Was `Mesh::GetTypicalElementTransformation` need just to use it in a call to `GetRule`?

pazner replied:
`GetTypicalElementTransformation` is needed mostly (only?) for getting the quadrature rules.

I think it's better if the "user code" to get the quadrature rules doesn't have conditionals based on empty partitions. If the user needs to branch based on `ne == 0`, it's very easy to forget to do this, and then the bug is only caught if the code is actually run with empty partitions.

We could also change the `GetRule` interface (or, for backwards compatibility, add a new interface and deprecate the old one) that doesn't take an `ElementTransformation`, and instead takes some other information (geometry, order of the Jacobian determinant, etc.) that fulfills the same purpose.

v-dobrev replied:
I'm beginning to think that the whole thing with the `GetTypical*` methods (that I started) is probably not the best approach to handle empty partitions. These calls to `GetTypical*` are only used because methods assume the a single type of element and order are present in the mesh. Thinking ahead, we want to generalize this to mixed meshes and variable orders, so these `GetTypical*` methods are just temporary addition that will become unnecessary in the future. Maybe we should try to deal with empty partitions in the context of dealing with mixed meshes and variable orders? What do you think?

pazner replied:
Yes, I think that the "GetTypical" approach is definitely a temporary solution that will not be necessary in the future.

I imagine that in the future, there will be some division of the elements into "classes" based on geometry, finite element polynomial degree, etc., and then all the elements of a certain class will be treated together in a batch. The code will have to handle potentially arbitrary numbers of classes, and so this will treat the empty partition case automatically (since in that case, there will just be zero classes).

Implementing that is probably a significant effort, and I don't know if there are concrete plans to do it currently. I think it's probably better to have a temporary "bandaid" fix for the empty partitions for now, and then refactor when the more general solution is implemented. Maybe we can mark the new "GetTypical" APIs as "internal" somehow, so that users know not to depend on them long-term.

v-dobrev replied:
Okay sounds reasonable. As a solution of the above change in signature of `GetRule` -- do you mind keeping the original version (for consistency, backward compatibility, and maybe future use) and just add the new version as a separate overload?

pazner replied:
Done in d8df6b802045a8149c02a803429f79dee56dc348

v-dobrev commented on fem/gridfunc.cpp:
```diff
@@ -1790,8 +1768,8 @@ void GridFunction::ProjectGridFunction(const GridFunction &src)
-      fes->GetFE(0)->Project(*src.fes->GetFE(0),
-                             *mesh->GetElementTransformation(0), P);
+      fes->GetTypicalFE()->Project(*src.fes->GetTypicalFE(),
+                                   *mesh->GetTypicalElementTransformation(), P);
```
Here we don't need to use any of the `GetTypical*` methods since the check `mesh->GetNumGeometries(mesh->Dimension()) == 1` guarantees that we have at least one element locally. And even before that, we have the early exit for empty meshes: `if (!mesh->GetNE()) { return; }`.

pazner replied:
True, but it's not clear to me that `GetFE(0)` is clearer (in terms of communicating intent) than `GetTypicalFE()`. The result will be the same. I'm OK with either one.

v-dobrev replied:
I guess my main point here was that we don't actually need `GetTypicalElementTransformation` here -- this is one place where `GetTypicalElementTransformation` is not used in conjunction with `GetRule`. Since we are not going to remove `GetTypicalElementTransformation`, there's no need to revert the change, if you prefer to keep it this way -- as you said the behavior is the same.

v-dobrev commented on fem/kdtree.cpp:
```diff
@@ -176,8 +176,8 @@ void KDTreeNodalProjection<2>::Project(const GridFunction& gf, real_t lerr)
       // initialize bbmax and bbmin
-      const FiniteElement* el=gf.FESpace()->GetFE(0);
-      trans = gf.FESpace()->GetElementTransformation(0);
+      const FiniteElement* el=gf.FESpace()->GetTypicalFE();
+      trans = gf.FESpace()->GetMesh()->GetTypicalElementTransformation();
```
First, I don't think this code is meant to run on the local portion of a parallel mesh, so assuming we have elements is probably just fine. @bslazarov, is this correct?

If this is supposed to work for empty local mesh (maybe not now but in the future), I'm not sure using some dummy `ElementTransformation` is the right thing to use here (if `Mesh::GetTypicalElementTransformation()` returns a dummy `ElementTransformation` in this case). I suspect that this case needs to be handled explicitly here.

pazner replied:
I think you're right, reverted in 06f60e28d.

v-dobrev commented on miniapps/tools/tmop-check-metric.cpp:
```diff
@@ -153,8 +153,8 @@ int main(int argc, char *argv[])
 
-   ElementTransformation &Tr = *mesh->GetElementTransformation(0);
-   const FiniteElement &fe = *fespace.GetFE(0);
+   ElementTransformation &Tr = *mesh->GetTypicalElementTransformation();
+   const FiniteElement &fe = *fespace.GetTypicalFE();
```
There's no need to change the code here -- the mesh with 1 element is constructed just above.

pazner replied:
Reverted in 4c96ebc65280284531ef470aa895753f061b3b2f.

tzanio commented:
@pazner, can you please resolve conflicts?

tzanio reviewed (approved):
Thanks @pazner !

sebastiangrimberg commented on mesh/mesh.cpp:
```diff
@@ -1486,6 +1525,34 @@ Array<int> Mesh::GetFaceToBdrElMap() const
+   {
+      return Geometry::SEGMENT;
+   }
+   const int mesh_gen = MeshGenerator();
```
Sorry for the drive by comment here, but why not just use the `meshgen` member here like elsewhere in the class? Either way I think its access and naming should be made consistent so we can ctrl-f in the library for either `meshgen` or `mesh_gen`.

pazner replied:
@v-dobrev I think you wrote this part. What do you think about using `meshgen` directly instead of defining the new local variable `mesh_gen`?

v-dobrev replied:
I don't remember why I wrote it this way -- maybe originally I had this code outside of the `Mesh` class, so I had to access `meshgen` it with the method. I don't see a downside to using `meshgen` directly, if you prefer it that way.

pazner replied:
Done in f88486fda17c095408467c1e61860806c93c7d2c

v-dobrev commented on fem/kdtree.cpp:
```diff
@@ -177,7 +177,7 @@ void KDTreeNodalProjection<2>::Project(const GridFunction& gf, real_t lerr)
       // initialize bbmax and bbmin
       const FiniteElement* el=gf.FESpace()->GetFE(0);
-      trans = gf.FESpace()->GetElementTransformation(0);
+      trans = gf.FESpace()->GetMesh()->GetElementTransformation(0);
```
This change seems unnecessary.

pazner replied:
Reverted in d85c8fb65a2914390083ac6252711b74602fb04e.

v-dobrev commented on fem/kdtree.cpp:
```diff
@@ -303,7 +303,7 @@ void KDTreeNodalProjection<3>::Project(const GridFunction& gf, real_t lerr)
       // initialize bbmax and bbmin
       const FiniteElement* el=gf.FESpace()->GetFE(0);
-      trans = gf.FESpace()->GetElementTransformation(0);
+      trans = gf.FESpace()->GetMesh()->GetElementTransformation(0);
```
Same -- seems unnecessary.

pazner replied:
Reverted in d85c8fb65a2914390083ac6252711b74602fb04e.

v-dobrev commented on tests/benchmarks/bench_assembly_levels.cpp:
```diff
@@ -260,7 +260,7 @@ struct Kernel: public BakeOff
 
    Kernel(AssemblyLevel assembly, int order, int N)
-   : BakeOff(assembly,order,N,VDIM,GLL), y(&fes)
+      : BakeOff(assembly,order,N,VDIM,GLL), y(&fes)
```
Looks like we need to add `tests/benchmarks` to the directories where `make style` is applied.

v-dobrev replied:
Actually there a few more sub-directories in `tests` that we should add: `convergence`, `mem_manager`, `par-mesh-format`. Also one sub-directory inside `tests/unit`: `enzyme`.

pazner replied:
Good catch, I added those directories in 697c21aa7e4a25f2ce7df09698c81e1fca86547a.

v-dobrev commented:
Overall this PR looks good to me. Thanks Will!

v-dobrev commented:
I think it makes sense to address #4554 here. I think we can do that by adding a call to `ReduceMeshGen()` just before the following line: https://github.com/mfem/mfem/blob/9a9087e20b214cfb74f684429bd912e21b9381b3/mesh/submesh/psubmesh.cpp#L114

It looks like at this point the rank-local `meshgen` is set and after the newly added `ReduceMeshGen()` it will no longer be modified.

pazner commented:
> I think it makes sense to address #4554 here. I think we can do that by adding a call to `ReduceMeshGen()` just before the following line:
> 
> https://github.com/mfem/mfem/blob/9a9087e20b214cfb74f684429bd912e21b9381b3/mesh/submesh/psubmesh.cpp#L114
> 
> It looks like at this point the rank-local `meshgen` is set and after the newly added `ReduceMeshGen()` it will no longer be modified.

Sounds good to me, I added the call to `ReduceMeshGen` in 78e593bea12c1ac85241b757ef25020dfc4ae4f0.

I also addressed the other review comments.

tzanio commented:
Merged in `next` for testing...
